beacon/light/committee_chain.go

```diff
@@ -0,0 +1,587 @@
+	updates            *canonicalStore[*types.LightClientUpdate]
+	committees         *canonicalStore[*types.SerializedSyncCommittee]
+	fixedRoots         *canonicalStore[common.Hash]
+	syncCommitteeCache *lru.Cache[uint64, syncCommittee] // cache deserialized committees
+
+	clock       mclock.Clock         // monotonic clock (simulated clock in tests)
+	unixNano    func() int64         // system clock (simulated clock in tests)
+	sigVerifier committeeSigVerifier // BLS sig verifier (dummy verifier in tests)
+
+	config             *types.ChainConfig
+	signerThreshold    int
+	minimumUpdateScore types.UpdateScore
+	enforceTime        bool
+}
+
+// NewCommitteeChain creates a new CommitteeChain.
+func NewCommitteeChain(db ethdb.KeyValueStore, config *types.ChainConfig, signerThreshold int, enforceTime bool, sigVerifier committeeSigVerifier, clock mclock.Clock, unixNano func() int64) *CommitteeChain {
+	s := &CommitteeChain{
+		fixedRoots: newCanonicalStore[common.Hash](db, rawdb.FixedRootKey, func(root common.Hash) ([]byte, error) {
+			return root[:], nil
+		}, func(enc []byte) (root common.Hash, err error) {
+			if len(enc) != common.HashLength {
+				return common.Hash{}, errors.New("incorrect length for committee root entry in the database")
+			}
+			return common.BytesToHash(enc), nil
+		}),
+		committees: newCanonicalStore[*types.SerializedSyncCommittee](db, rawdb.SyncCommitteeKey, func(committee *types.SerializedSyncCommittee) ([]byte, error) {
+			return committee[:], nil
+		}, func(enc []byte) (*types.SerializedSyncCommittee, error) {
+			if len(enc) == types.SerializedSyncCommitteeSize {
+				committee := new(types.SerializedSyncCommittee)
+				copy(committee[:], enc)
+				return committee, nil
+			}
+			return nil, errors.New("incorrect length for serialized committee entry in the database")
+		}),
+		updates: newCanonicalStore[*types.LightClientUpdate](db, rawdb.BestUpdateKey, func(update *types.LightClientUpdate) ([]byte, error) {
+			return rlp.EncodeToBytes(update)
+		}, func(enc []byte) (*types.LightClientUpdate, error) {
+			update := new(types.LightClientUpdate)
+			if err := rlp.DecodeBytes(enc, update); err != nil {
+				return nil, err
+			}
+			return update, nil
+		}),
+		syncCommitteeCache: lru.NewCache[uint64, syncCommittee](10),
+		db:                 db,
+		sigVerifier:        sigVerifier,
+		clock:              clock,
+		unixNano:           unixNano,
+		config:             config,
+		signerThreshold:    signerThreshold,
+		enforceTime:        enforceTime,
+		minimumUpdateScore: types.UpdateScore{
+			SignerCount:    uint32(signerThreshold),
+			SubPeriodIndex: params.SyncPeriodLength / 16,
+		},
+	}
+
+	// check validity constraints
+	if !s.updates.periods.IsEmpty() {
+		if s.fixedRoots.periods.IsEmpty() || s.updates.periods.First < s.fixedRoots.periods.First ||
+			s.updates.periods.First >= s.fixedRoots.periods.AfterLast {
+			log.Error("Inconsistent database error: first update is not in the fixed roots range")
```

Comment:
Originally I used `log.Crit` so that the client quits and the user can manually reset the chain. Martin suggested to just print errors but you are right, now the client just keeps running with an invalid database. Now I changed it so that it resets the chain automatically on startup if the database is invalid.

Reply:
In general, I think things like `cmd/xxx` can do `log.Crit`, but most other parts that may potentially be used as libraries, should _not_ use `log.Crit`. `log.Crit` takes the decision "nothing to do we need to kill the process immediately". And a library should not do that, it should return with an error, and let the caller decide if it needs to die immediately or if it can do things like close files and databases first. 

I usually complain when people use `log.Crit`, since I consider it a bit "the lazy way out"

Reply:
Makes sense. So here we can either refuse to create `CommitteeChain` or just reset it (I chose the latter as it allows recovering by resyncing).